### DISCUSSION TITLE
Is the voxel order deterministic in voxel_grid ?

### DISCUSSION THREAD

**Discussion by Wang-Tianyu** [pyg-team/pytorch_geometric#4481] [Q&A]
I am wondering, does the voxel_grid function index the set of voxels according to a deterministic order?
Or equivalently, if I apply the voxel grid with the same start, end, and size parameter on different point clouds, will I get consistent voxel indices?

Is there a way to quickly index voxel vertex and get their positions?

**Comment by rusty1s** [comment]
The voxel computation is deterministic. Sadly, there does not exist an elegant way to get voxel centers right now, but this shouldn’t be too hard to add. Let me know if you have interest in adding this :)

**Reply by Wang-Tianyu** [reply]
After some investigation, I found some rather unusual behavior in the implementation of the voxel grid function with torch_cluster 1.6.0
1. The start argument sets the location of the voxel of index 0, but all points with coordinates smaller than start will still be assigned to voxels but with possibly negative index.
2. Not only that, if we set voxel grid start to be at [x, 0, z] and voxel size to be 1,  then points of coordinate [a, y, c], where y \in [-1, 1), will be assigned the same voxel index value. (yes, there can be a voxel that is 8 times larger).
3. While the doc says that the end indicates the "End coordinates of the grid"  I think it actually does nothing.
Since all points with coordinates larger than end will also be assigned a voxel index.
4. If one point happens to have at least one coordinate to be the same as the end, and the start - end is divisible by size, then that point may be assigned to a voxel that is located outside the volume defined by start and end.

Of course, none of the above will happen (or at least not likely) if you let the function figure out the start and end itself, which is the most common use case.
Below I provide a notebook that showcases the above-mentioned problem in a visualized way.
Plus, in the notebook, I show a way to return the voxel vertex index and position given the voxel index, with some conditions.

It would be good to do all that within the CUDA code but I am not familiar with CUDA coding.

[voxel_grid_test.zip](https://github.com/pyg-team/pytorch_geometric/files/8505298/voxel_grid_test.zip)
